vm/rust/src/jsonrpc.rs

```diff
@@ -5,30 +5,80 @@ use starknet_api::hash::StarkFelt;
 use starknet_api::transaction::{Calldata, EventContent, EthAddress, L2ToL1Payload};
 use starknet_api::deprecated_contract_class::EntryPointType;
 use blockifier::execution::entry_point::CallType;
+use starknet_api::transaction::{Transaction as StarknetApiTransaction};
 
 #[derive(Serialize)]
-pub struct TransactionTrace {
-    pub validate_invocation: Option<FunctionInvocation>,
-    pub execute_invocation: Option<FunctionInvocation>,
-    pub fee_transfer_invocation: Option<FunctionInvocation>,
+#[serde(untagged)]
+pub enum TransactionTrace {
+    // used for INVOKE_TXN_TRACE and DECLARE_TXN_TRACE
+    Common {
+        #[serde(skip_serializing_if = "Option::is_none")]
+        validate_invocation: Option<FunctionInvocation>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        execute_invocation: Option<FunctionInvocation>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        fee_transfer_invocation: Option<FunctionInvocation>,
+    },
+    DeployAccount {
+        #[serde(skip_serializing_if = "Option::is_none")]
+        validate_invocation: Option<FunctionInvocation>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        constructor_invocation: Option<FunctionInvocation>,
+        #[serde(skip_serializing_if = "Option::is_none")]
+        fee_transfer_invocation: Option<FunctionInvocation>,
+    },
+    L1Handler {
+        #[serde(skip_serializing_if = "Option::is_none")]
+        function_invocation: Option<FunctionInvocation>
+    }
```

Comment:
```suggestion
        #[serde(skip_serializing_if = "Option::is_none")]
        validate_invocation: Option<FunctionInvocation>,
        #[serde(skip_serializing_if = "Option::is_none")]
        execute_invocation: Option<FunctionInvocation>,
        #[serde(skip_serializing_if = "Option::is_none")]
        constructor_invocation: Option<FunctionInvocation>,
        #[serde(skip_serializing_if = "Option::is_none")]
        fee_transfer_invocation: Option<FunctionInvocation>,
        #[serde(skip_serializing_if = "Option::is_none")]
        function_invocation: Option<FunctionInvocation>
```

Results in less verbose code IMO, feel free to disregard tho